### PR TITLE
Add tests that verify elliptic solutions

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

These tests verify that the solutions indeed numerically solve the first-order system by applying the elliptic operator.

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [ ] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/PointwiseFunctions/AnalyticSolutions/Poisson/Lorentzian.cpp [hunk 1]
```diff
@@ -22,6 +23,21 @@ tuples::TaggedTuple<Tags::Field> Lorentzian<3>::variables(
...
 }
 
+template <>
+tuples::TaggedTuple<
+    ::Tags::deriv<Tags::Field, tmpl::size_t<3>, Frame::Inertial>>
+Lorentzian<3>::variables(
+    const tnsr::I<DataVector, 3>& x,
+    tmpl::list<::Tags::deriv<Tags::Field, tmpl::size_t<3>,
+                             Frame::Inertial>> /*meta*/) noexcept {
+  const DataVector prefactor = -1. / cube(sqrt(1 + get(dot_product(x, x))));
```
Comment (carmaza) [inline src/PointwiseFunctions/AnalyticSolutions/Poisson/Lorentzian.cpp]: `sqrt(1. + ` for consistency? (The `Tags::Field` `variables` above would need this change too.)

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/FirstOrderEllipticSolutionsTestHelpers.hpp [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+#include "Utilities/MakeWithValue.hpp"
+#include "tests/Unit/TestHelpers.hpp"
+#include "tests/Utilities/MakeWithRandomValues.hpp"
+
+namespace FirstOrderEllipticSolutionsTestHelpers {
+
+namespace Tags {
+
+template <typename Tag>
+struct OperatorAppliedTo : db::SimpleTag, db::PrefixTag {
```
Comment (carmaza) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/FirstOrderEllipticSolutionsTestHelpers.hpp:28]: I believe that (ideally) you want to `#include` the corresponding `db` headers. This comment also applies to the `db` functions used in `vefiry_solution`.

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/FirstOrderEllipticSolutionsTestHelpers.hpp [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+#include "Domain/LogicalCoordinates.hpp"
+#include "Domain/Mesh.hpp"
+#include "Elliptic/FirstOrderComputeTags.hpp"
+#include "Elliptic/FirstOrderOperator.hpp"
+#include "NumericalAlgorithms/LinearOperators/Divergence.tpp"
+#include "Utilities/MakeWithValue.hpp"
+#include "tests/Unit/TestHelpers.hpp"
+#include "tests/Utilities/MakeWithRandomValues.hpp"
+
+namespace FirstOrderEllipticSolutionsTestHelpers {
```
Comment (carmaza) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/FirstOrderEllipticSolutionsTestHelpers.hpp:23]: If you're not planning on putting anything else in this file, I'd probably call it `VerifyFirstOrderEllipticSolution.hpp`, get rid of this namespace, and put the `Tags` below in a `detail` namespace, in order to sort of keep consistency with what's done in `GrMhd` and `Burgers`.

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/FirstOrderEllipticSolutionsTestHelpers.hpp [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+template <size_t Dim>
+struct InverseJacobian : db::SimpleTag {
+  using type =
+      ::InverseJacobian<DataVector, Dim, Frame::Logical, Frame::Inertial>;
+};
+
+}  // namespace Tags
+
+/// Tests that the `solution` numerically solves the `System` on the given grid
```
Comment (carmaza) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/FirstOrderEllipticSolutionsTestHelpers.hpp]: ...for the given tolerance.

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/Poisson/Test_Lorentzian.cpp [hunk 5]
```diff
@@ -70,5 +85,13 @@ SPECTRE_TEST_CASE(
...
-  test_solution<3>();
+  using AffineMap = domain::CoordinateMaps::Affine;
+  using AffineMap3D =
+      domain::CoordinateMaps::ProductOf3Maps<AffineMap, AffineMap, AffineMap>;
+  test_solution(
+      Mesh<3>{12, Spectral::Basis::Legendre,
+              Spectral::Quadrature::GaussLobatto},
+      domain::CoordinateMap<Frame::Logical, Frame::Inertial, AffineMap3D>{
+          {{-1., 1., -0.5, 0.5}, {-1., 1., -0.5, 0.5}, {-1., 1., -0.5, 0.5}}},
+      1.e-1);
```
Comment (carmaza) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/Poisson/Test_Lorentzian.cpp]: For my own understanding: what criterion are you using for choosing different tolerances for different solutions and different dimensions?

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/FirstOrderEllipticSolutionsTestHelpers.hpp (hunk 3 above)
Comment (nilsvu) [reply tests/Unit/PointwiseFunctions/AnalyticSolutions/FirstOrderEllipticSolutionsTestHelpers.hpp:23]: I'm not 100% sure about the namespace and where to place the file yet.. perhaps other reviewers can chime in?
- It's possible that I'll want to add more test helpers for elliptic analytic solutions in the future
- I prefer to put these test helper functions in a namespace. Else I'm always confused where these free functions are coming from until I find the header file that includes them.

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/Poisson/Test_Lorentzian.cpp (hunk 5 above)
Comment (nilsvu) [reply tests/Unit/PointwiseFunctions/AnalyticSolutions/Poisson/Test_Lorentzian.cpp]: Your comment made me go back and do this more thoroughly. In the _"fixup tolerances for smooth and non-smooth functions"_ commit I add tests that the error is not only within a given tolerance but also _decreases_ as expected. For smooth functions we expect exponential convergence, and for non-smooth functions power-law convergence when increasing the polynomial order. The parameters of the convergence (e.g. the offset and scaling of the exponential) depend on the problem and the domain, so I determined those experimentally. For example, these are a bunch of norms that show how the Lorentzian solution converges, and an exponential tolerance with C1=5 and C2=1.2:

Unknown-24

Please let me know if this makes things more clear!

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/FirstOrderEllipticSolutionsTestHelpers.hpp (hunk 3 above)
Comment (carmaza) [reply tests/Unit/PointwiseFunctions/AnalyticSolutions/FirstOrderEllipticSolutionsTestHelpers.hpp:23]: Fair enough.

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/FirstOrderEllipticSolutionsTestHelpers.hpp [hunk 6]
```diff
@@ -85,11 +91,114 @@ void verify_solution(
...
+        const auto& fixed_source =
+            get<::Tags::FixedSource<field_tag>>(fixed_sources);
+        double l2_error_square = 0.;
+        double linf_error = 0.;
+        for (size_t i = 0; i < operator_applied_to_field.size(); ++i) {
+          const auto error = operator_applied_to_field[i] - fixed_source[i];
+          l2_error_square += alg::accumulate(square(error), 0.0) / error.size();
+          const double component_linf_error =
+              *std::max_element(error.begin(), error.end());
+          if (component_linf_error > linf_error) {
```
Comment (carmaza) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/FirstOrderEllipticSolutionsTestHelpers.hpp]: You might need to take `abs(error)` somewhere before making this comparison.

**Review by carmaza — COMMENTED** [review]
OK, thanks for explaining the choice of tolerances. The new version of the test looks neat.

I don't have any other comments, so feel free to squash.

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/FirstOrderEllipticSolutionsTestHelpers.hpp (hunk 6 above)
Comment (nilsvu) [reply tests/Unit/PointwiseFunctions/AnalyticSolutions/FirstOrderEllipticSolutionsTestHelpers.hpp]: good catch, thanks! squashed this in as well

**Comment by nilsvu** [comment]
@sxs-collaboration/spectre-core-devs This went through a round of review. Could you provide the second or assign another reviewer?

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/FirstOrderEllipticSolutionsTestHelpers.hpp [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+  using operator_applied_to_fields_tag =
+      db::add_tag_prefix<Tags::OperatorAppliedTo, fields_tag>;
+  using fixed_sources_tag = db::add_tag_prefix<::Tags::FixedSource, fields_tag>;
+
+  const auto logical_coords = logical_coordinates(mesh);
+  const auto inertial_coords = coord_map(logical_coords);
+  const auto solution_fields = variables_from_tagged_tuple(solution.variables(
+      inertial_coords, db::get_variables_tags_list<fields_tag>{}));
+
+  auto box = db::create<
```
Comment (kidder) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/FirstOrderEllipticSolutionsTestHelpers.hpp]: please make this test independent of `DataBox`.  This is the style of programming that we want to discourage.  When you know what everything is, just use free functions.

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/FirstOrderEllipticSolutionsTestHelpers.hpp [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+        const auto& fixed_source =
+            get<::Tags::FixedSource<field_tag>>(fixed_sources);
+        double l2_error_square = 0.;
+        double linf_error = 0.;
+        for (size_t i = 0; i < operator_applied_to_field.size(); ++i) {
+          const auto error =
+              abs(operator_applied_to_field[i] - fixed_source[i]);
+          l2_error_square += alg::accumulate(square(error), 0.0) / error.size();
+          const double component_linf_error =
+              *std::max_element(error.begin(), error.end());
```
Comment (kidder) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/FirstOrderEllipticSolutionsTestHelpers.hpp]: use `alg::max_element`

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/FirstOrderEllipticSolutionsTestHelpers.hpp [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+            sqrt(l2_error_square / operator_applied_to_field.size());
+        CAPTURE(db::tag_name<field_tag>());
+        CAPTURE(l2_error);
+        CAPTURE(linf_error);
+        CHECK(l2_error < tolerance);
+        CHECK(linf_error < tolerance);
+      });
+}
+
+/*!
```
Comment (kidder) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/FirstOrderEllipticSolutionsTestHelpers.hpp:109]: `\ingroup TestingFrameworkGroup` for both of these and add this file to the appropriate place in `docs/Doxyfile.in` for this documentation to be rendered

**Review by kidder — CHANGES_REQUESTED** [review]

**Comment by nilsdeppe** [comment]
@nilsleiffischer this is waiting on you.

**Comment by nilsvu** [comment]
@kidder I made this independent of DataBox compute items now. Please take a look at the commit _"Make elliptic first-order operator a free function"_ as well as the fixup.

@carmaza could you take a look at those new commits as well please?

**Comment by nilsvu** [comment]
@kidder @carmaza please take a look at this PR :)

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/FirstOrderEllipticSolutionsTestHelpers.hpp [hunk 10]
```diff
@@ -31,97 +30,79 @@ struct OperatorAppliedTo : db::SimpleTag, db::PrefixTag {
...
-        CHECK(linf_error < tolerance);
-      });
+  tmpl::for_each<all_fields>([&operator_applied_to_fields, &fixed_sources,
+                              &tolerance](auto field_tag_v) {
+    using field_tag = tmpl::type_from<decltype(field_tag_v)>;
+    const auto& operator_applied_to_field =
+        get<Tags::OperatorAppliedTo<field_tag>>(operator_applied_to_fields);
+    const auto& fixed_source =
+        get<::Tags::FixedSource<field_tag>>(fixed_sources);
+    double l2_error_square = 0.;
```
Comment (carmaza) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/FirstOrderEllipticSolutionsTestHelpers.hpp:89]: Either 0. or 0.0 in all occurrences.

**Review by carmaza — COMMENTED** [review]
No objections to the new changes on my side.

**Review by kidder — COMMENTED** [review]
I am happy with the rewritten test.  You currently have a linking error.  You can squash and rebase when fixing that.

**Comment by nilsdeppe** [comment]
Still broken.

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/Poisson/Test_Lorentzian.cpp [hunk 11]
```diff
@@ -9,10 +9,19 @@
 #include "DataStructures/DataBox/Prefixes.hpp"  // IWYU pragma: keep
 #include "DataStructures/DataVector.hpp"
 #include "DataStructures/Tensor/TypeAliases.hpp"
+#include "Domain/CoordinateMaps/Affine.hpp"
+#include "Domain/CoordinateMaps/CoordinateMap.hpp"
+#include "Domain/CoordinateMaps/CoordinateMap.tpp"
```
Comment (nilsvu) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/Poisson/Test_Lorentzian.cpp:14]: These `tpp` includes were needed after the rebase to fix the linking errors.

**Comment by nilsdeppe** [comment]
@carmaza @kidder This should be ready for you two to look at again :)

**Review by carmaza — APPROVED** [review]

**Comment by nilsvu** [comment]
@kidder please take a look

**Review by kidder — APPROVED** [review]

**Pull request merged**